ocaml/xapi/xapi_cluster.ml

```diff
@@ -100,17 +110,75 @@ let pool_create ~__context ~network ~cluster_stack ~token_timeout ~token_timeout
+let filter_on_option opn xs =
+  match opn with
+  | None -> xs
+  | Some x -> List.filter ((<>) x) xs
+
+(* Helper function; concurrency checks are done in implementation of Cluster.destroy and Cluster_host.destroy *)
+let pool_force_destroy ~__context ~self = 
+  (* For now we assume we have only one pool, and that the cluster is the same as the pool.
+     This means that the pool master must be a member of this cluster. *)
+  let master = Helpers.get_master ~__context in
+  let master_cluster_host =
+    Xapi_clustering.find_cluster_host ~__context ~host:master
+  in
+  let slave_cluster_hosts =
+    Db.Cluster.get_cluster_hosts ~__context ~self |> filter_on_option master_cluster_host
+  in
+  (* First try to destroy each cluster_host - if we can do so safely then do *)
+  List.iter
+    (fun cluster_host -> 
+      (* We need to run this code on the slave *)
+      (* We ignore failures here, we'll try a force_destroy after *)
+      log_and_ignore_exn (fun () ->  
+        Helpers.call_api_functions ~__context (fun rpc session_id ->
+            Client.Client.Cluster_host.destroy ~rpc ~session_id ~self:cluster_host)
+      )
+    )
+    slave_cluster_hosts;
+  (* We expect destroy to have failed for some, we'll try to force destroy those *)
+  (* Note we include the master here, we should attempt to force destroy it *)
+  let all_remaining_cluster_hosts =
+    Db.Cluster.get_cluster_hosts ~__context ~self
+  in
+  (* Now try to force_destroy, keep track of any errors here *)
+  let exns = List.fold_left
+    (fun exns_so_far cluster_host ->
+      Helpers.call_api_functions ~__context (fun rpc session_id ->
+        try
+          Client.Client.Cluster_host.force_destroy ~rpc ~session_id ~self:cluster_host;
+          exns_so_far
+        with e ->
+          Backtrace.is_important e;
+          let uuid = Client.Client.Cluster_host.get_uuid ~rpc ~session_id ~self:cluster_host in
+          debug "Ignoring exception while trying to force destroy cluster host %s: %s" uuid (ExnHelper.string_of_exn e);
+          e :: exns_so_far
+      )
+    )
+    [] all_remaining_cluster_hosts
+    in
+
+    begin
+    match exns with
+    | [] -> ()
+    | e :: _ -> raise Api_errors.(Server_error (cluster_force_destroy_failed, [ExnHelper.string_of_exn e]))
+    end;
+
+    Helpers.call_api_functions ~__context (fun rpc session_id ->
+        Client.Client.Cluster.destroy ~rpc ~session_id ~self)
+
 (* Helper function; concurrency checks are done in implementation of Cluster.destroy and Cluster_host.destroy *)
 let pool_destroy ~__context ~self =
   (* For now we assume we have only one pool, and that the cluster is the same as the pool.
      This means that the pool master must be a member of this cluster. *)
   let master = Helpers.get_master ~__context in
   let master_cluster_host =
     Xapi_clustering.find_cluster_host ~__context ~host:master
-    |> Xapi_stdext_monadic.Opt.unbox
```

Comment:
A while ago I tried another alternative change that makes `pool_destroy` more distributed: https://github.com/gaborigloi/xen-api/commit/b0bc29d6662f02e3d317810ab1bcaeab489f462d
This should work when the master has no `Cluster_host`. However, it didn't work because `Cluster.destroy` is always redirected to master. So we'd have to change message forwarding too.
What happens with this change if we destroy the cluster host of the master and the do a pool destroy on the cluster?

Reply:
We've discussed that this would fail if there is no master `Cluster_host` because in that case we'd try `Cluster_host.destroy` on the last one, and that would fail.

Reply:
As we discussed I'll remove my changes to `pool_destroy` and we'll use your patch later.